errors/errors.go

```diff
@@ -0,0 +1,110 @@
+package errors
+
+import "fmt"
+
+// ErrorType provides a coarse category for BoulderErrors
+type ErrorType int
+
+const (
+	InternalServer ErrorType = iota
+	NotSupported
+	Malformed
+	Unauthorized
+	NotFound
+	SignatureValidation
+	RateLimit
+	TooManyRequests
+	RejectedIdentifier
+	UnsupportedIdentifier
+	InvalidEmail
+	ConnectionFailure
+)
+
+// BoulderError represents internal Boulder errors
+type BoulderError struct {
+	Type   ErrorType
+	Detail string
+}
+
+func (be *BoulderError) Error() string {
+	return be.Detail // add a stringer for the ErrorType?
+}
+
+// New is a convenience function for creating a new BoulderError
+func New(errType ErrorType, msg string, args ...interface{}) error {
+	return &BoulderError{
+		Type:   errType,
+		Detail: fmt.Sprintf(msg, args...),
+	}
+}
+
+// Is is a convenience function for testing the internal type of an BoulderError
+func Is(err error, errType ErrorType) bool {
+	bErr, ok := err.(*BoulderError)
+	if !ok {
+		return false
+	}
+	return bErr.Type == errType
+}
+
+// ChangeType is a convenience function for switching the internal type of a BoulderError.
+// If the passed error is not a BoulderError it will simply be wrapped with the requested
+// type
+func ChangeType(err error, errType ErrorType) error {
+	bErr, ok := err.(*BoulderError)
+	if !ok {
+		return New(errType, err.Error())
+	}
+	bErr.Type = errType
+	return bErr
+}
+
+// These methods should all be generated programmatically
```

Comment:
I think it might be overkill based on how often we add errors, but maybe I'm underestimating the frequency we'll encounter this. Maybe it's a good idea to track the idea as a feature-req issue and we can come back around to it when we know more about whether it will be a common enough event to justify the overhead.